### DISCUSSION TITLE
Allow usage of iframes with src=domains listed in HTML Field Security

### DISCUSSION THREAD

**Discussion by erikthomte** [microsoft-search/pnp-modern-search#1966] [Ideas]
**Is your feature request related to a problem? Please describe.**
First of all a big kudos to Franky , Wobba and the rest of the team for developing and maintaining PnP Modern Search !
I have recently used v 4.5.4 in a project, and enjoyed it very much.

One thing has been an issue; the area for preview image/interactive preview is very small. I managed to partially solve this by using the component "Panel", and use a large version of the preview image inside the panel. But for some msg/eml files the preview image doesn't work well, even with large size in Panel settings. And only the first page shows in preview image (std. SP functionality I know)

A better/larger interactive preview would be very nice.
This can probably only be acomplished by using iframes, and using built in SP functionality, aka Previewdemo.aspx or wopiframe, "open in browser" etc.

I understand that iframes are a potential security risk, but SharePoint itself uses iframes for document viewer (internal links) and SharePoint also has the embedded webpart that accepts iframe content from all domains in Site Settings - "HTML Field Security".

Using this site setting also allows site admins/owners to control which domains that will be whitelisted as source for iframes.
PnP Modern Web parts shouldn't need stricter rules that other built in webparts in SharePoint, IHMO.

**Describe the solution you'd like**
Allow usage of iframes for all search results templates, for all domains listed in HTML Field Security.
eg something like.
DOMPurify.sanitize(dirty, { ADD_TAGS: ["iframe"], ADD_ATTR: ['allow', 'allowfullscreen', 'frameborder', 'scrolling', 'height', 'width', 'style'] });
Support for handlebars would be nice (example: switch iframe src reference based on filtype)

**Describe alternatives you've considered**
Panel solution described above, but it has no interactivity, and doesn't work well with some msg/eml files

**Additional context**

**Comment by wobba** [comment]
I don't necessarily disagree and would welcome a PR for this. First you need a setting to allow iframes which would then perform the API call per page load to read the setting of allowed domains. 

Another and maybe better approach is to create a custom web component with the logic you need, and use this component in your handlebar template.

**Comment by erikthomte** [comment]
Thanks for the quick reply @wobba . Unfortunately I'm not really a developer (20% developer, 30% advanced superuser and 50% "I'll fix it somehow") and I don't have any skills in SharePoint Framework, so a custom web component is out of my reach. But I do have a lot of "workaround" ideas :grinning:

If the question/context was instead "Allow usage of iframes with src=current domain" - would that be considered as an possible enchantment for the standard web parts in PnP ?

Idea: Clone Modern PnP Search, and make changes only to DomPurifyHelper.ts  - Add isCurrentDomain logic and add dompurify hooks to allow for iframe tags with src=current domain. Recompile (with the help of an colleague) and use.
This idea would depend on how dompurify is implemented in the PnP webparts. Is this idea worth considering?

Another idea (if I can recruit a colleague) : To reduce the amount of work needed - only make a custom web component based on PnP Panel (not changing every search results web part). Add new params -  data-autoload="true/false", and data-autoloadURL="<any url>". Is there any potential issues/gotchas with this approach?

Last one: For the pnp-filepreview builtn web component - would it be possible to use handlebar replace function to increase width and height, and/or add some css so the preview would be larger? I assume this is not possible based on earlier comments from Franky, but I just had to ask

**Reply by wobba** [reply]
Anything is possible as long as someone implements it :) Feel free to log each idea as a separate discussion point for someone to pick up if they think it's important to get in.

**Comment by erikthomte** [comment]
In the spirit of "sharing is caring" and giving back to this community I thought I would share my "Large but not interactive" preview solution. The large preview image shows in the pnp-panel content area, and is a clickable link to full preview in new tab.

```html
<pnp-panel   
    data-is-open="false" 
    data-is-light-dismiss="true"
    data-is-blocking="true"
    data-size="5"
    data-panel-header-text="Preview {{Filename}}">
    <template id="panel-open">
         <!-- All the content here will be wrapped with an onclick event opening/hiding the panel -->         
       <img src="/sites/resources/pnpsearch/icons/search64x64.png" width="16" height="16" alt="Preview for: {{Filename}}">
    </template>
    <template id="panel-content">
        <!-- Panel content goes here -->
    <a href="{{slot item @root.slots.ParentLink}}/?id={{replace (slot item @root.slots.DocumentLink) "https://contoso.sharepoint.com" ""}}&parent={{replace (slot item @root.slots.ParentLink) "https://contoso.sharepoint.com" ""}}" data-interception="off" title="Preview for: {{Filename}}" target="_blank">
    <img src="{{AutoPreviewImageUrl}}" alt="Preview for: {{Filename}}" loading="lazy" style="border:2px solid #a2a2a2; border-radius: 4px; padding: 2px">
    </a>
    </template>
</pnp-panel>
```

**Comment by mrpullen** [comment]
So I would like to use this to surface a powerapp interface on the pnp-panel - assuming I write a custom web component that simply takes parameters and the appId - any other advice? 

Is this something that could go into the PnP Search as a supported component?

**Reply by wobba** [reply]
@mrpullen sure... it's basically an iframe with the app id, so feel free to create a component for it. Saves people from generating the iframe URL :)
